### PR TITLE
Add inertia-free approach

### PR DESCRIPTION
**Pull request by nychiang**
Add inertia-free approach to HiOp.
Only the dWd curvature test is implemented. (see paper [An Inertia-free Filter Line-search Algorithm For Large-scale Nonlinear Programming](https://link.springer.com/article/10.1007/s10589-015-9820-y)
The current inertia-free test only applies to option the KKT linear system with option `xdycyd` or `xycyd'

Add one test problem `nlpSparse_Ex7_2` to test inertia-free approach.

@pelesh @cnpetra

### CODE REVIEW

File: src/Drivers/nlpSparse_ex7_driver.cpp [hunk 1]
```diff
@@ -115,23 +146,28 @@ int main(int argc, char **argv)
...
 
 
-static bool self_check(size_type n, double objval)
+static bool self_check(size_type n, double objval, const bool inertia_free)
 {
 #define num_n_saved 3 //keep this is sync with n_saved and objval_saved
   const size_type n_saved[] = {50, 500, 5000};
-  const double objval_saved[] = { -1.58349999995100e+03, -1.53428124950100e+03, -1.04209374500105e+03};
+  double objval_saved[] = { -1.58349999995100e+03, -1.53428124950100e+03, -1.04209374500105e+03};
+  if(inertia_free) {
```
Comment (cnpetra) [inline src/Drivers/nlpSparse_ex7_driver.cpp:154]: we need to discuss this next time we meet.

File: src/Optimization/hiopKKTLinSysDense.hpp [hunk 2]
```diff
@@ -242,7 +242,7 @@ class hiopKKTLinSysDenseXDYcYd : public hiopKKTLinSysCompressedXDYcYd
   * [    Jc        0     0      0    ] [dyc] = [   ryc    ]
   * [    Jd       -I     0      0    ] [dyd]   [   ryd    ]
   */
-  virtual bool updateMatrix( const double& delta_wx, const double& delta_wd,
+  virtual bool build_kkt_matrix( const double& delta_wx, const double& delta_wd,
```
Comment (cnpetra) [inline src/Optimization/hiopKKTLinSysDense.hpp:245]: alignment

File: src/Optimization/hiopKKTLinSysMDS.cpp [hunk 3]
```diff
@@ -168,7 +168,7 @@ namespace hiop
   }
 
 
-  bool hiopKKTLinSysCompressedMDSXYcYd::updateMatrix(const double& delta_wx, const double& delta_wd,
+  bool hiopKKTLinSysCompressedMDSXYcYd::build_kkt_matrix(const double& delta_wx, const double& delta_wd,
                                                      const double& delta_cc, const double& delta_cd)
   {
```
Comment (cnpetra) [inline src/Optimization/hiopKKTLinSysMDS.cpp:173]: alignment

**Comment by nychiang** [comment]
@cnpetra I have addressed all the issues.

**Comment by cnpetra** [comment]
is this ready for review?

**Comment by nychiang** [comment]
yes

File: src/Drivers/nlpSparse_ex7.cpp [hunk 4]
```diff
@@ -7,14 +7,14 @@
 /** Nonlinear *highly nonconvex* and *rank deficient* problem test for the Filter IPM
  * Newton of HiOp. It uses a Sparse NLP formulation. The problem is based on Ex6.
  *
- *  min   -(2*convex_obj-1)*sum 1/4* { (x_{i}-1)^4 : i=1,...,n} + 0.5x^Tx
+ *  min   (2*convex_obj-1)*scal*sum 1/4* { (x_{i}-1)^4 : i=1,...,n} + 0.5x^Tx
```
Comment (cnpetra) [inline src/Drivers/nlpSparse_ex7.cpp:10]: please document `scal`. Also not sure the variable name `scal_neg_obj` used later on is the best one. Looks like a better name would be `scale_quartic_obj_term` ...

File: src/Optimization/hiopFactAcceptor.cpp [hunk 5]
```diff
@@ -4,7 +4,7 @@
 //
 // This file is part of HiOp. For details, see https://github.com/LLNL/hiop. HiOp 
 // is released under the BSD 3-clause license (https://opensource.org/licenses/BSD-3-Clause). 
-// Please also read �Additional BSD Notice� below.
+// Please also read �Additional BSD Notice� below.
```
Comment (cnpetra) [inline src/Optimization/hiopFactAcceptor.cpp:7]: there are some strange characters enclosing "Additional BSD Notice", please use enclosing quotes

File: src/Optimization/hiopKKTLinSysSparse.cpp [hunk 6]
```diff
@@ -640,7 +640,7 @@ namespace hiop
     return dynamic_cast<hiopLinSolverNonSymSparse*> (linSys_);
   }
 
-  bool hiopKKTLinSysSparseFull::updateMatrix(const double& delta_wx, const double& delta_wd,
+  bool hiopKKTLinSysSparseFull::build_kkt_matrix(const double& delta_wx, const double& delta_wd,
```
Comment (cnpetra) [inline src/Optimization/hiopKKTLinSysSparse.cpp:643]: alignment

File: src/Utils/hiopOptions.cpp [hunk 7]
```diff
@@ -818,7 +818,14 @@ void hiopOptionsNLP::register_options()
...
                         range,
                         "The criteria used to accept a factorization: inertia_correction (default option) "
-                        " inertia_free (to be supported soon)");
+                        "and inertia_free.");
+    register_num_option("neg_curv_test_fact",
+                        1e-11,
+                        0.,
+                        1e+20,
+                        "Apply curvature test to check if a factorization is acceptable. "
+                        "This is the scaling factor which used to determines if the "
```
Comment (cnpetra) [inline src/Utils/hiopOptions.cpp:827]: remove "which"

**Review by cnpetra — CHANGES_REQUESTED** [review]
looks great, just left a couple of small observations

File: src/Optimization/hiopKKTLinSysMDS.hpp [hunk 8]
```diff
@@ -107,7 +107,7 @@ class hiopKKTLinSysCompressedMDSXYcYd : public hiopKKTLinSysCompressedXYcYd
 		      const hiopMatrix* Jac_c, const hiopMatrix* Jac_d,
 		      hiopMatrix* Hess);
 
-  virtual bool updateMatrix(const double& delta_wx, const double& delta_wd,
+  virtual bool build_kkt_matrix(const double& delta_wx, const double& delta_wd,
                             const double& delta_cc, const double& delta_cd);
```
[the remark this answers is not in the document]
Comment (cnpetra) [reply src/Optimization/hiopKKTLinSysMDS.hpp:114]: this does not seem to be fixed: fix alignment and codestyle

File: src/Optimization/hiopKKTLinSysSparse.cpp [hunk 9]
```diff
@@ -81,7 +81,7 @@ namespace hiop
     delete Hx_;
   }
 
-  bool hiopKKTLinSysCompressedSparseXYcYd::updateMatrix(const double& delta_wx, const double& delta_wd,
+  bool hiopKKTLinSysCompressedSparseXYcYd::build_kkt_matrix(const double& delta_wx, const double& delta_wd,
```
Comment (cnpetra) [inline src/Optimization/hiopKKTLinSysSparse.cpp]: this does not seem to be fixed: fix alignment and codestyle

File: src/Optimization/hiopKKTLinSysSparse.cpp [hunk 10]
```diff
@@ -348,7 +348,7 @@ namespace hiop
     delete Hd_;
   }
 
-  bool hiopKKTLinSysCompressedSparseXDYcYd::updateMatrix(const double& delta_wx, const double& delta_wd,
+  bool hiopKKTLinSysCompressedSparseXDYcYd::build_kkt_matrix(const double& delta_wx, const double& delta_wd,
```
Comment (cnpetra) [inline src/Optimization/hiopKKTLinSysSparse.cpp]: this does not seem to be fixed: fix alignment and codestyle

File: src/Optimization/hiopKKTLinSysSparse.cpp (hunk 6 above)
Comment (cnpetra) [reply src/Optimization/hiopKKTLinSysSparse.cpp]: this does not seem to be fixed: fix alignment and codestyle

File: src/Optimization/hiopKKTLinSysSparse.hpp [hunk 11]
```diff
@@ -78,7 +78,7 @@ class hiopKKTLinSysCompressedSparseXYcYd : public hiopKKTLinSysCompressedXYcYd
   hiopKKTLinSysCompressedSparseXYcYd(hiopNlpFormulation* nlp);
   virtual ~hiopKKTLinSysCompressedSparseXYcYd();
 
-  virtual bool updateMatrix(const double& delta_wx, const double& delta_wd,
+  virtual bool build_kkt_matrix(const double& delta_wx, const double& delta_wd,
                             const double& delta_cc, const double& delta_cd);
```
[the remark this answers is not in the document]
Comment (cnpetra) [reply src/Optimization/hiopKKTLinSysSparse.hpp]: this does not seem to be fixed: fix alignment and codestyle

File: src/Optimization/hiopKKTLinSysSparse.hpp [hunk 12]
```diff
@@ -138,7 +138,7 @@ class hiopKKTLinSysCompressedSparseXDYcYd : public hiopKKTLinSysCompressedXDYcYd
   hiopKKTLinSysCompressedSparseXDYcYd(hiopNlpFormulation* nlp);
   virtual ~hiopKKTLinSysCompressedSparseXDYcYd();
 
-  virtual bool updateMatrix(const double& delta_wx, const double& delta_wd,
+  virtual bool build_kkt_matrix(const double& delta_wx, const double& delta_wd,
                             const double& delta_cc, const double& delta_cd);
```
[the remark this answers is not in the document]
Comment (cnpetra) [reply src/Optimization/hiopKKTLinSysSparse.hpp]: this does not seem to be fixed: fix alignment and codestyle

File: src/Optimization/hiopKKTLinSysSparse.hpp [hunk 13]
```diff
@@ -209,7 +209,7 @@ class hiopKKTLinSysSparseFull : public hiopKKTLinSysFull
 
   virtual ~hiopKKTLinSysSparseFull();
 
-  virtual bool updateMatrix(const double& delta_wx, const double& delta_wd,
+  virtual bool build_kkt_matrix(const double& delta_wx, const double& delta_wd,
                             const double& delta_cc, const double& delta_cd);
```
[the remark this answers is not in the document]
Comment (cnpetra) [reply src/Optimization/hiopKKTLinSysSparse.hpp]: this does not seem to be fixed: fix alignment and codestyle

File: src/Utils/hiopOptions.cpp (hunk 7 above)
Comment (cnpetra) [reply src/Utils/hiopOptions.cpp]: not removed

File: src/Optimization/hiopFactAcceptor.cpp (hunk 5 above)
Comment (cnpetra) [reply src/Optimization/hiopFactAcceptor.cpp]: not fixed

**Review by cnpetra — APPROVED** [review]
apparently github does not show the mods correctly. I checked manually and all the issues appear fixed.

**Pull request merged**